simulation/amaru-sim/src/simulator/simulate.rs

```diff
@@ -262,11 +291,105 @@ mod tests {
         assert_eq!(world.run_world(), &Vec::new());
     }
 
-    // NOTE: This will be replaced with a test that uses an in-process NodeHandle later, which
-    // avoids the problem of locating and spawning the binary.
+    #[test]
+    #[should_panic]
+    fn simulate_pure_stage_echo() {
+        let config = Default::default();
+
+        let number_of_nodes = 1;
+
+        let spawn: fn() -> NodeHandle = || {
+            println!("*** Spawning node!");
+            let mut network = SimulationBuilder::default();
+            let stage = network.stage(
+                "echo",
+                async |(mut state, out), msg: Envelope<EchoMessage>, eff| {
+                    if let EchoMessage::Echo { msg_id, echo } = &msg.body {
+                        state += 1;
+                        // Insert a bug every 5 messages.
+                        let reply;
+                        if state % 5 == 0 {
+                            reply = Envelope {
+                                src: msg.dest,
+                                dest: msg.src,
+                                body: EchoMessage::EchoOk {
+                                    msg_id: state,
+                                    in_reply_to: *msg_id,
+                                    echo: echo.to_string().to_uppercase(),
+                                },
+                            };
+                        } else {
+                            reply = Envelope {
+                                src: msg.dest,
+                                dest: msg.src,
+                                body: EchoMessage::EchoOk {
+                                    msg_id: state,
+                                    in_reply_to: *msg_id,
+                                    echo: echo.to_string(),
+                                },
+                            };
+                        };
+                        println!(" ==> {:?}", reply);
+                        eff.send(&out, reply).await;
+                        Ok((state, out))
+                    } else {
+                        panic!("Got a message that wasn't an echo: {:?}", msg.body)
+                    }
+                },
+                (0u64, StageRef::noop::<Envelope<EchoMessage>>()),
+            );
+            let (output, rx) = network.output("output");
+            let stage = network.wire_up(stage, |state| state.1 = output.without_state());
+            let running = network.run();
+
+            pure_stage_node_handle(rx, stage, running).unwrap()
+        };
+        let generate_message = (0..128u8).prop_map(|i| EchoMessage::Echo {
+            msg_id: 0,
+            echo: format!("Please echo {}", i),
+        });
+        let property = |trace: Trace| {
+            for msg in trace.0.iter().filter(|msg| msg.src.starts_with("c")) {
+                if let EchoMessage::Echo { msg_id, echo } = &msg.body {
+                    let response = trace.0.iter().find(|resp| {
```

Comment:
if we were serious about this test, we’d want to start looking only after the echo request and stop looking after a certain time has passed in the trace